File: api/rh_api/services/process_flow.py
from __future__ import annotations

from .helpers import normalize_compare_text, normalize_text


PROCESS_STATUS_CLOSED = "Encerrado"

CANDIDATE_STATUS_ANALYSIS = "Analise"
CANDIDATE_STATUS_QUALIFIED = "Qualificado"
CANDIDATE_STATUS_NOT_QUALIFIED = "Nao qualificado"
CANDIDATE_STATUS_SCHEDULED = "Agendado"
CANDIDATE_STATUS_CONFIRMED = "Confirmado"
CANDIDATE_STATUS_ATTENDED = "Compareceu"
CANDIDATE_STATUS_MISSED = "Faltou"
CANDIDATE_STATUS_APPROVED = "Aprovado"
CANDIDATE_STATUS_ELIMINATED = "Eliminado"
CANDIDATE_STATUS_TALENT_BANK = "Banco de talentos"

INTERVIEW_OPERATIONAL_STATUSES = {
    CANDIDATE_STATUS_SCHEDULED,
    CANDIDATE_STATUS_CONFIRMED,
    CANDIDATE_STATUS_ATTENDED,
    CANDIDATE_STATUS_MISSED,
}

TERMINAL_CANDIDATE_STATUSES = {
    CANDIDATE_STATUS_APPROVED,
    CANDIDATE_STATUS_ELIMINATED,
    CANDIDATE_STATUS_TALENT_BANK,
}

FINAL_DECISION_ALLOWED_STATUSES = {
    CANDIDATE_STATUS_ANALYSIS,
    CANDIDATE_STATUS_ATTENDED,
}

INTERVIEW_SCHEDULING_ALLOWED_STATUSES = {
    CANDIDATE_STATUS_QUALIFIED,
}


def normalize_process_status(status: str | None) -> str:
    safe_status = normalize_compare_text(status)
    if safe_status == "encerrado":
        return PROCESS_STATUS_CLOSED
    return normalize_text(status)


def is_process_closed(status: str | None) -> bool:
    return normalize_process_status(status) == PROCESS_STATUS_CLOSED


def canonicalize_candidate_status(status: str | None) -> str:
    safe_status = normalize_compare_text(status)
    if not safe_status:
        return CANDIDATE_STATUS_ANALYSIS

    if safe_status in {"analise", "em analise"}:
        return CANDIDATE_STATUS_ANALYSIS
    if safe_status == "qualificado":
        return CANDIDATE_STATUS_QUALIFIED
    if safe_status == "nao qualificado":
        return CANDIDATE_STATUS_NOT_QUALIFIED
    if safe_status == "agendado":
        return CANDIDATE_STATUS_SCHEDULED
    if safe_status == "confirmado":
        return CANDIDATE_STATUS_CONFIRMED
    if safe_status == "compareceu":
        return CANDIDATE_STATUS_ATTENDED
    if safe_status == "faltou":
        return CANDIDATE_STATUS_MISSED
    if safe_status == "aprovado":
        return CANDIDATE_STATUS_APPROVED
    if safe_status == "banco de talentos":
        return CANDIDATE_STATUS_TALENT_BANK
    if safe_status == "reprovado" or "eliminado" in safe_status:
        return CANDIDATE_STATUS_ELIMINATED

    return normalize_text(status)


def get_candidate_visible_status(
    status_candidato: str | None,
    status_entrevista: str | None = None,
) -> str:
    candidate_status = canonicalize_candidate_status(status_candidato)
    interview_status = canonicalize_candidate_status(status_entrevista)

    if candidate_status in TERMINAL_CANDIDATE_STATUSES:
        return candidate_status
    if interview_status in INTERVIEW_OPERATIONAL_STATUSES:
        return interview_status
    return candidate_status


def is_terminal_candidate_status(status: str | None) -> bool:
    return canonicalize_candidate_status(status) in TERMINAL_CANDIDATE_STATUSES


def status_allows_final_decision(status: str | None) -> bool:
    return canonicalize_candidate_status(status) in FINAL_DECISION_ALLOWED_STATUSES


def status_allows_interview_scheduling(status: str | None) -> bool:
    return canonicalize_candidate_status(status) in INTERVIEW_SCHEDULING_ALLOWED_STATUSES


def map_cv_classification_to_status(classification: str | None) -> str:
    safe_classification = normalize_compare_text(classification)
    if safe_classification == "qualificado":
        return CANDIDATE_STATUS_QUALIFIED
    return CANDIDATE_STATUS_NOT_QUALIFIED


def build_process_closed_message(action_label: str, id_processo: str | None = None) -> str:
    process_suffix = f" no processo {normalize_text(id_processo)}" if normalize_text(id_processo) else ""
    return f"O processo seletivo esta encerrado{process_suffix}. Nao e permitido {action_label}."


def build_candidate_status_action_label(status: str | None) -> str:
    safe_status = canonicalize_candidate_status(status)

    if safe_status == CANDIDATE_STATUS_APPROVED:
        return "aprovar o candidato"
    if safe_status == CANDIDATE_STATUS_ELIMINATED:
        return "eliminar o candidato"
    if safe_status == CANDIDATE_STATUS_TALENT_BANK:
        return "enviar o candidato para banco de talentos"
    if safe_status in INTERVIEW_OPERATIONAL_STATUSES:
        return "atualizar o status operacional do candidato"

    return "movimentar o candidato"

File: Front/fonte/shared/process-flow.js
const PROCESS_STATUS_CLOSED = 'Encerrado';

const CANDIDATE_STATUS_ANALYSIS = 'Analise';
const CANDIDATE_STATUS_QUALIFIED = 'Qualificado';
const CANDIDATE_STATUS_NOT_QUALIFIED = 'Nao qualificado';
const CANDIDATE_STATUS_SCHEDULED = 'Agendado';
const CANDIDATE_STATUS_CONFIRMED = 'Confirmado';
const CANDIDATE_STATUS_RESCHEDULED = 'Reagendado';
const CANDIDATE_STATUS_ATTENDED = 'Compareceu';
const CANDIDATE_STATUS_MISSED = 'Faltou';
const CANDIDATE_STATUS_APPROVED = 'Aprovado';
const CANDIDATE_STATUS_ELIMINATED = 'Eliminado';
const CANDIDATE_STATUS_TALENT_BANK = 'Banco de talentos';

const INTERVIEW_STATUSES = new Set([
  CANDIDATE_STATUS_SCHEDULED,
  CANDIDATE_STATUS_CONFIRMED,
  CANDIDATE_STATUS_RESCHEDULED,
  CANDIDATE_STATUS_ATTENDED,
  CANDIDATE_STATUS_MISSED,
]);

const TERMINAL_STATUSES = new Set([
  CANDIDATE_STATUS_NOT_QUALIFIED,
  CANDIDATE_STATUS_APPROVED,
  CANDIDATE_STATUS_ELIMINATED,
  CANDIDATE_STATUS_TALENT_BANK,
]);

function normalizeCompareText(value) {
  return String(value || '')
    .trim()
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '');
}

export function normalizeProcessStatus(status) {
  return normalizeCompareText(status) === 'encerrado'
    ? PROCESS_STATUS_CLOSED
    : String(status || '').trim();
}

export function isProcessClosed(statusOrProcess) {
  if (typeof statusOrProcess === 'object' && statusOrProcess !== null) {
    return normalizeProcessStatus(
      statusOrProcess.status || statusOrProcess.status_processo,
    ) === PROCESS_STATUS_CLOSED;
  }

  return normalizeProcessStatus(statusOrProcess) === PROCESS_STATUS_CLOSED;
}

export function canonicalizeCandidateStatus(status) {
  const value = normalizeCompareText(status);

  if (!value) return CANDIDATE_STATUS_ANALYSIS;
  if (value === 'analise' || value === 'em analise') return CANDIDATE_STATUS_ANALYSIS;
  if (value === 'qualificado') return CANDIDATE_STATUS_QUALIFIED;
  if (value === 'nao qualificado') return CANDIDATE_STATUS_NOT_QUALIFIED;
  if (value === 'agendado') return CANDIDATE_STATUS_SCHEDULED;
  if (value === 'entrevista agendada') return CANDIDATE_STATUS_SCHEDULED;
  if (value === 'confirmado') return CANDIDATE_STATUS_CONFIRMED;
  if (value === 'reagendado') return CANDIDATE_STATUS_RESCHEDULED;
  if (value === 'compareceu') return CANDIDATE_STATUS_ATTENDED;
  if (value === 'faltou') return CANDIDATE_STATUS_MISSED;
  if (value === 'aprovado') return CANDIDATE_STATUS_APPROVED;
  if (value === 'banco de talentos') return CANDIDATE_STATUS_TALENT_BANK;
  if (value === 'reprovado' || value.includes('eliminado')) {
    return CANDIDATE_STATUS_ELIMINATED;
  }

  return String(status || '').trim();
}

export function getCandidateVisibleStatus(candidateOrStatus, statusEntrevista = '') {
  if (
    candidateOrStatus &&
    typeof candidateOrStatus === 'object' &&
    !Array.isArray(candidateOrStatus)
  ) {
    if (candidateOrStatus.status_fluxo) {
      return canonicalizeCandidateStatus(candidateOrStatus.status_fluxo);
    }

    return getCandidateVisibleStatus(
      candidateOrStatus.status_candidato,
      candidateOrStatus.status_entrevista,
    );
  }

  const candidateStatus = canonicalizeCandidateStatus(candidateOrStatus);
  const interviewStatus = String(statusEntrevista || '').trim()
    ? canonicalizeCandidateStatus(statusEntrevista)
    : '';

  if (TERMINAL_STATUSES.has(candidateStatus)) {
    return candidateStatus;
  }

  if (INTERVIEW_STATUSES.has(interviewStatus)) {
    return interviewStatus;
  }

  return candidateStatus;
}

export function isTerminalCandidateStatus(status) {
  return TERMINAL_STATUSES.has(canonicalizeCandidateStatus(status));
}

export function getCandidateActionState(candidate, processStatus = '') {
  const visibleStatus = getCandidateVisibleStatus(candidate);
  const closed = isProcessClosed(processStatus || candidate?.status_processo);
  const schedule = !closed && visibleStatus === CANDIDATE_STATUS_QUALIFIED;
  const finalDecision =
    !closed &&
    (visibleStatus === CANDIDATE_STATUS_ANALYSIS ||
      visibleStatus === CANDIDATE_STATUS_ATTENDED);

  return {
    visibleStatus,
    processClosed: closed,
    isFinalized:
      closed ||
      isTerminalCandidateStatus(visibleStatus) ||
      visibleStatus === CANDIDATE_STATUS_MISSED,
    canScheduleInterview: schedule,
    canApprove: finalDecision,
    canEliminate: finalDecision,
    canSendToTalentBank: finalDecision,
  };
}

export function getCandidateFlowGroup(candidate) {
  const status = getCandidateVisibleStatus(candidate);

  if (status === CANDIDATE_STATUS_QUALIFIED) {
    return 'Qualificacao';
  }

  if (INTERVIEW_STATUSES.has(status)) {
    return 'Entrevista';
  }

  if (TERMINAL_STATUSES.has(status)) {
    return 'Finalizado';
  }

  return 'Analise';
}

export function getPipelineStageLabel(stage) {
  const value = String(stage || '').trim();

  if (value === 'Triagem') return 'Analise';
  if (value === 'Prova') return 'Qualificados';
  if (value === 'Entrevista') return 'Entrevistas';
  if (value === 'Aprovado') return 'Aprovados';
  if (value === 'Reprovado') return 'Finalizados';
  return value || 'Etapa';
}

export {
  CANDIDATE_STATUS_ANALYSIS,
  CANDIDATE_STATUS_APPROVED,
  CANDIDATE_STATUS_ATTENDED,
  CANDIDATE_STATUS_CONFIRMED,
  CANDIDATE_STATUS_ELIMINATED,
  CANDIDATE_STATUS_MISSED,
  CANDIDATE_STATUS_NOT_QUALIFIED,
  CANDIDATE_STATUS_QUALIFIED,
  CANDIDATE_STATUS_RESCHEDULED,
  CANDIDATE_STATUS_SCHEDULED,
  CANDIDATE_STATUS_TALENT_BANK,
  PROCESS_STATUS_CLOSED,
};
